server/vulnerabilities/nvd/cpe.go

```diff
@@ -297,6 +297,48 @@ func CPEFromSoftware(db *sqlx.DB, software *fleet.Software, translations CPETran
 	return "", nil
 }
 
+func consumeCPEBuffer(
+	ctx context.Context,
+	ds fleet.Datastore,
+	logger kitlog.Logger,
+	batch []fleet.SoftwareCPE,
+) error {
+	var toDelete []fleet.SoftwareCPE
+	var toUpsert []fleet.SoftwareCPE
+
+	for i := range batch {
+		// This could be because of a new translation rule or because we fixed a bug with the CPE
+		// detection process
+		if batch[i].CPE == "" {
+			toDelete = append(toDelete, batch[i])
+			continue
+		}
+		toUpsert = append(toUpsert, batch[i])
+	}
+
+	if len(toUpsert) != 0 {
+		upserted, err := ds.UpsertSoftwareCPEs(ctx, toUpsert)
+		if err != nil {
+			return err
+		}
+		if int(upserted) != len(toUpsert) {
+			level.Debug(logger).Log("toUpsert", len(toUpsert), "upserted", upserted)
```

Comment:
Anything we typically have to do where we have such an error?

Reply:
+1. Is there anything to do if `int(upserted) != len(toUpsert)`? Maybe we don't need `UpsertSoftwareCPEs` to return the rows affected.

Reply:
So one scenario where `int(upserted) != len(toUpsert)` is if there are duplicate values in `toUpsert` in that case then the end result might not be what we want depending on the order in which the `INSERT ...` statements are executed. I can't see how this scenario would be possible with the current implementation, but this could change in the future.